[BUG] enum classes cannot be used in cdef class method declarations

### Describe the bug

Cython generates invalid C++ code when an enum class is used as either the return type or a parameter of a method of a cdef class. The issue is that the `__PYX_ENUM_CLASS_DECL` macro is inlined too late in the file to be available at the time when it is first needed for a declaration.

### Code to reproduce the behaviour:

```cython
# distutils: language = c++


cdef enum class my_enum(int):
    a = 1
    b = 0


cdef class A:
    cdef my_enum f(self, my_enum x):
        return x
```

When I attempt to compile this code, I get:
```
(main) dt08% cythonize -3 -i -f definitions.pyx
[1/1] Cythonizing /home/nfs/vyasr/local/testing/cython_enum_class_decl/definitions.pyx
/home/nfs/vyasr/local/testing/cython_enum_class_decl/definitions.cpp:1404:3: error: ‘__PYX_ENUM_CLASS_DECL’ does not name a type
 1404 |   __PYX_ENUM_CLASS_DECL __pyx_t_11definitions_my_enum (*f)(struct __pyx_obj_11definitions_A *, __PYX_ENUM_CLASS_DECL __pyx_t_11definitions_my_enum);
      |   ^~~~~~~~~~~~~~~~~~~~~
/home/nfs/vyasr/local/testing/cython_enum_class_decl/definitions.cpp: In function ‘int __Pyx_modinit_type_init_code()’:
/home/nfs/vyasr/local/testing/cython_enum_class_decl/definitions.cpp:3512:32: error: ‘struct __pyx_vtabstruct_11definitions_A’ has no member named ‘f’
 3512 |   __pyx_vtable_11definitions_A.f = (__PYX_ENUM_CLASS_DECL __pyx_t_11definitions_my_enum (*)(struct __pyx_obj_11definitions_A *, __PYX_ENUM_CLASS_DECL __pyx_t_11definitions_my_enum))__pyx_f_11definitions_1A_f;
      |                                ^
error: command '/usr/bin/gcc' failed with exit code 1
```

If I inspect definitions.cpp, the `__PYX_ENUM_CLASS_DECL` macro is in fact declared in the file, but it is declared after the first usage, causing this issue.

### Expected behaviour

I expect this code to compile successfully.

### OS

Linux

### Python version

3.11.4

### Cython version

3.0.0

### Additional context

_No response_

--- a/Cython/Utility/CppSupport.cpp
+++ b/Cython/Utility/CppSupport.cpp
@@ -69,6 +69,7 @@ auto __Pyx_pythran_to_python(T &&value) -> decltype(to_python(
 #endif
 
 ////////////// EnumClassDecl.proto //////////////////
+//@proto_block: utility_code_proto_before_types
 
 #if defined (_MSC_VER)
   #if _MSC_VER >= 1910

--- a/tests/compile/cppenum.pyx
+++ b/tests/compile/cppenum.pyx
@@ -40,3 +40,9 @@ cdef enum class Color(int):
 cdef enum class Color2(int):
     RED = (<int> Color.RED)
     GREEN = (<int> Color.GREEN)
+
+
+# enum class as cdef class function parameter
+cdef class A:
+    cdef Spam f(self, Spam s):
+        return s

